src/main/java/net/neoforged/neoforge/fluids/CauldronFluidContent.java

```diff
@@ -0,0 +1,169 @@
+ * If {@link #levelProperty} is {@code null}, then {@code maxLevel = 1}, and there is only one level.
+ * Otherwise, the levels are all the integer values between {@code 1} and {@link #maxLevel} (included).
+ * </li>
+ * <li>{@link #totalAmount} defines how much fluid (in millibuckets) there is in one level of the cauldron.</li>
+ * </ul>
+ */
+public final class CauldronFluidContent {
+    /**
+     * Block of the cauldron.
+     */
+    public final Block block;
+    /**
+     * Fluid stored inside the cauldron.
+     */
+    public final Fluid fluid;
+    /**
+     * Amount of {@code #fluid} in millibuckets in the entire full cauldron.
+     */
+    public final int totalAmount;
+    /**
+     * Maximum level for {@link #levelProperty}. {@code 1} if {@code levelProperty} is null, otherwise a number {@code >= 1}.
+     * The minimum level is always 1.
+     */
+    public final int maxLevel;
+    /**
+     * Property storing the level of the cauldron. If it's {@code null}, only one level is possible.
+     */
+    @Nullable
+    public final IntegerProperty levelProperty;
+
+    /**
+     * Return the current level of the cauldron given its block state, or 0 if it's an empty cauldron.
+     */
+    public int currentLevel(BlockState state) {
+        if (fluid == Fluids.EMPTY) {
+            return 0;
+        } else if (levelProperty == null) {
+            return 1;
+        } else {
+            return state.getValue(levelProperty);
+        }
+    }
+
+    private CauldronFluidContent(Block block, Fluid fluid, int totalAmount, int maxLevel, @Nullable IntegerProperty levelProperty) {
+        this.block = block;
+        this.fluid = fluid;
+        this.totalAmount = totalAmount;
+        this.maxLevel = maxLevel;
+        this.levelProperty = levelProperty;
+    }
+
+    private static final Map<Block, CauldronFluidContent> BLOCK_TO_CAULDRON = new IdentityHashMap<>();
+    private static final Map<Fluid, CauldronFluidContent> FLUID_TO_CAULDRON = new IdentityHashMap<>();
+
+    /**
+     * Get the cauldron fluid content for a cauldron block, or {@code null} if none was registered (yet).
+     */
+    @Nullable
+    public static CauldronFluidContent getForBlock(Block block) {
+        return BLOCK_TO_CAULDRON.get(block);
+    }
+
+    /**
+     * Get the cauldron fluid content for a fluid, or {@code null} if no cauldron was registered for that fluid (yet).
+     */
+    @Nullable
+    public static CauldronFluidContent getForFluid(Fluid fluid) {
+        return FLUID_TO_CAULDRON.get(fluid);
+    }
+
+    @ApiStatus.Internal
+    public static void init() {
+        var registerEvent = new RegisterCauldronFluidContentEvent();
+        // Vanilla registrations
+        registerEvent.register(Blocks.CAULDRON, Fluids.EMPTY, 1000, null);
+        registerEvent.register(Blocks.WATER_CAULDRON, Fluids.WATER, 1000, LayeredCauldronBlock.LEVEL);
+        registerEvent.register(Blocks.LAVA_CAULDRON, Fluids.LAVA, 1000, null);
```

Comment:
Do we want these to use `FluidType.BUCKET_VOLUME` instead of just hardcoding `1000` in another place?

Reply:
I knew the constant was somewhere just didn't find it :smile: